### ISSUE TITLE
help about .dialog

### ISSUE THREAD

**Issue by EduModsS** [commonuserlol/frida-java-menu#22]
help me, how to make this dialog work, its doing anything..

```function main() {

	const AssemblyCSharp = Il2Cpp.domain.assembly("Assembly-CSharp").image;

	const CharacterMotor = AssemblyCSharp.class("SYBO.RunnerCore.Character.CharacterMotor");

	const get_CanJump = CharacterMotor.method("get_CanJump");

	let menu = new Menu.JavaMenu("your title", "your subtitle", new Menu.LGLLayout());

	get_CanJump.implementation = function() {
		if (infJump === false) {
			return this.method("get_CanJump").invoke();
		} else {
			return true;
		}
	}

	const layout = new Menu.LGLLayout();
	const composer = new Menu.Composer("your title", "your subtitle", layout);
	composer.icon("https://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/skull.png", "Web");

	let infJump = false;

	Menu.add(layout.button("Pulo Infinitooo", function () {
		//menu.dialog("Title", "Cool message", () => toast("this will be shown when user press positive button"), () => toast("this will be shown when user press negative button"));
		Menu.dialog("Title", "Cool message", () => toast("this will be shown when user press positive button"), () => toast("this will be shown when user press negative button"));
		if (infJump === false) {
			infJump = true;
		} else {
			infJump = false;
		}
		this.allCaps = true;
	}));

	composer.show();
}```

**Comment by commonuserlol** [comment]
Hey, please look at wiki, I wrote some examples here (many changes was since your code was ok). Even better if your editor have typescript support

**Comment by EduModsS** [comment]
I had already looked on the wiki, but I didn't find any examples of how to use the functions there

**Comment by commonuserlol** [comment]
Ah my bad, you can create dialog via 
```typescript
const dialog = await layout.dialog("title", "message", {
        "fn": () => console.log("ok"),
        "label": "OK"
    },
    {
        "fn": () => console.log("fail"),
        "label": "EXIT"
    });
```

and then show it:
```typescript
dialog.show();
```

Note: if your function isn't async, use something like:
```typescript
layout.dialog("title", "message", {
        "fn": () => console.log("ok"),
        "label": "OK"
    },
    {
        "fn": () => console.log("fail"),
        "label": "EXIT"
    }).then((d) => d.show());
```

**Comment by EduModsS** [comment]
Could you give an example too, of input text and input number?

**Comment by commonuserlol** [comment]
```typescript
layout.inputNumber("title", 1337, {
    "fn": (value: number) => console.log(value),
    "label": "OK",
     },
     {
    "fn": () => console.log("fail"),
    "label": "EXIT"
});
```
inputText almost same, it have no max value (1337 number above) but have hint argument at end (callback value argument have `string` type)

**Comment by EduModsS** [comment]
it just shows the m m, then it freezes... maybe missing something?

async function main() {
	console.log("  m m ");
	const dialog = await layout.dialog("Welcome", "Welcome to mod !", {
		"fn": () => console.log("ok"),
		"label": "OK"
		},
	{
		"fn": () => console.log("fail"),
		"label": "EXIT"
	}
	);
	console.log("  m m mm ");

**Comment by EduModsS** [comment]
even without await, it freezes

**Comment by commonuserlol** [comment]
What freezes? More info pls

**Comment by commonuserlol** [comment]
can you show your code?

**Comment by EduModsS** [comment]
I used npx esbuild, and I'm editing there, I found it easier haha, here's the code, it's only giving errors on inputNumber and dialog, the rest is working perfectly :

```async function main() {

	const AssemblyCSharp = Il2Cpp.domain.assembly("Assembly-CSharp").image;

	const CharacterMotor = AssemblyCSharp.class("SYBO.RunnerCore.Character.CharacterMotor");

	const get_CanJump = CharacterMotor.method("get_CanJump");

	get_CanJump.implementation = function() {
		if (infJump === false) {
			return this.method("get_CanJump").invoke();
		} else {
			return true;
		}
	}

	const layout = new Menu.LGLLayout();
	const composer = new Menu.Composer("your title", "your subtitle", layout);
	//composer.icon("https://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/skull.png", "Web");
	composer.icon("base64blabla", "Normal");

	let infJump = false;

	let [layoutForAdd, layoutForElements] = layout.collapse("teste", false);

	let x = layout.inputNumber("title", 1337, {
		"fn": function(){console.log(value) },
		"label": "OK",
	},
	{
		"fn": function(){console.log(value) },
		"label": "EXIT"
	});

	x.then((xx) => {
		console.log(xx);
	}

	Menu.add(layout.toggle("MOD", function () {}), layoutForElements);
	Menu.add(layout.category("BragaMods"), layoutForElements);

	Menu.add(layoutForAdd);
	layout.category("BragaMods");
	// Toggle (nome / função)
	Menu.add(layout.button("Pulo Infinitooo", function () {
		if (infJump === false) {
			this.text = "BRAGA GAY";
			this.backgroundColor = "#000000";
			Menu.toast("Pulo Infinito Ativado", 0);
			infJump = true;
		} else {
			this.text = "KIRA GAY";
			this.backgroundColor = "#FFFFFF";
			Menu.toast("Pulo Infinito Desativado", 1);
			infJump = false;
		}
	}));
	Menu.add(layout.category("BragaMods2"));

	let seekval = 'a'

	// Seekbar (Nome / maximo / minimo / função)
	Menu.add(layout.seekbar(`nome {0}`, 9, 0, function(val) {
		
	}));

	// Spinner ( array / 
	Menu.add(layout.spinner(["função 1", "função 2", "função 3", "função 4"], function(val) {
		
	}));

	Menu.add(
		layout.radioGroup(
			"Choice: {0}",
			["1 button", "second button", "3rd button"],
			function(index) {
				switch (index) {
					case 0:
						infJump = true;
						Menu.toast("1", 0);
						break;
					case 1:
						infJump = false;
						Menu.toast("2", 0);
						break;
					case 2:
						Menu.toast("3", 0);
						break;
				}
			}
		)
	);

	composer.show();
}```

**Comment by EduModsS** [comment]
oh and also, I'm trying it both on frida gadget and on both frida normal (server + termux  frida installation)

**Comment by commonuserlol** [comment]
Probably main activity not available yet, can you create `input*` not in main function, in callback or smth

**Comment by EduModsS** [comment]
Could you give an example?

**Comment by EduModsS** [comment]
also, I'm calling main like this, so I don't think you can put it before the way it is:

Il2Cpp.perform(() => {
	Menu.waitForInit(main);
});

**Comment by EduModsS** [comment]
Isn't this something related to view ? 
async dialog(title, message, positiveCallback, negativeCallback, view) {...

**Comment by commonuserlol** [comment]
> Il2Cpp.perform

please read https://github.com/commonuserlol/frida-java-menu/releases/tag/v2.0.1 release notes

**Comment by EduModsS** [comment]
> > Il2Cpp.perform
> 
> leia https://github.com/commonuserlol/frida-java-menu/releases/tag/v2.0.1 notas de versão

I didn't understand exactly what you meant in this post

**Comment by commonuserlol** [comment]
TLDR: don't use Il2Cpp.perform with .waitForInit
Use il2cpp.perform ONLY inside your hooks/callbacks

**Comment by EduModsS** [comment]
I was a bit busy, I tried it in another game and it didn't work either, I remember I had used an old version of yours, I don't remember which one worked, maybe I'm writing something wrong or Not supporting Android 11? , Could you take a look at my code and tell me something that might help me fix it?

**Comment by EduModsS** [comment]
[FRIDAJAVAMENU.txt](https://github.com/commonuserlol/frida-java-menu/files/14622538/FRIDAJAVAMENU.txt)

**Comment by EduModsS** [comment]
when I do x as using await, it waits infinitely

**Comment by commonuserlol** [comment]
Async function as callback won't work, consider removing async/await for .waitForInit callback

**Comment by EduModsS** [comment]
I removed it, I just left main() ... however, the console.log never appeared
	let x = layout.inputNumber("title", 1337, {
		"fn": function(){},
		"label": "OK",
	},
	{
		"fn": function(){},
		"label": "EXIT"
	});

	x.then( (xx) => {
		console.log('input Number Loaded');
	});

**Comment by EduModsS** [comment]
.../📁 - Frida Nodejs/🟢 PvZ Heroes $ frida -H localhost -f com.kiloo.subwaysurf -l FRIDAJAVAMENU.js
     ____
    / _  |   Frida 16.2.1 - A world-class dynamic instrumentation toolkit
   | (_| |
    > _  |   Commands:
   /_/ |_|       help      -> Displays the help system
   . . . .       object?   -> Display information about 'object'
   . . . .       exit/quit -> Exit
   . . . .
   . . . .   More info at https://frida.re/docs/home/
   . . . .
   . . . .   Connected to localhost (id=socket@localhost)
Spawned `com.kiloo.subwaysurf`. Resuming main thread!
[Remote::com.kiloo.subwaysurf ]-> Menu already showed, ignoring `show` call
[Remote::com.kiloo.subwaysurf ]->

**Comment by commonuserlol** [comment]
do you put dialog ctor into callback? in main it may be ignored

**Comment by EduModsS** [comment]
how?

like

Menu.waitForInit( () => {
    const layout = new Menu.LGLLayout();

    	let x = layout.inputNumber("title", 1337, {
		"fn": function(){},
		"label": "OK",
	},
	{
		"fn": function(){},
		"label": "EXIT"
	});

	x.then((xx) => {
		console.log(xx);
	});
});


like this?? if no, how?

**Comment by commonuserlol** [comment]
try put it into button callback or smth

**Comment by EduModsS** [comment]
can you give example, inside a button callback?

**Comment by commonuserlol** [comment]
```typescript
Menu.add(layout.button("test", () => { // your dialog code here }));
```

**Comment by EduModsS** [comment]
in the callback i put the 

```
layout.inputNumber("title", 1337, {
	"fn": function(){},
	"label": "OK",
},
{
	"fn": function(){},
	"label": "EXIT"
});
```


?

**Comment by EduModsS** [comment]
os instead of layout i use anything different?

**Comment by commonuserlol** [comment]
yeah put that code which u sent

**Comment by EduModsS** [comment]
even inside the button it didn't worked, bad thing, better i ignore this problem lol, the rest is working normally

**Comment by commonuserlol** [comment]
you didn't called .show in dialog

**Comment by EduModsS** [comment]
i called x.then((xx) => { console.log(" loaded"); }); 
and no consolelog appeared

**Comment by commonuserlol** [comment]
can you don't assign dialog to variable, and directly call `then`?

**Comment by EduModsS** [comment]
how to use, the .dialog?

**Comment by commonuserlol** [comment]
just do `Menu.dialog(...).then((x) => x.show());` in your button callback

**Comment by commonuserlol** [comment]
or what you use, inputNumber

**Comment by EduModsS** [comment]
I tried this, and it didn't work:

```
	Menu.add(layout.button("TTTTT", () => { // your dialog code here }));
		layout.inputNumber(
			"title",
			1337, {
				"fn": function(){},
				"label": "OK"
			}, {
				"fn": function(){},
				"label": "EXIT"
			}
		).then((x) => {
			x.show();
		});
	}));
```

**Comment by EduModsS** [comment]
ah 😂, I just saw the error lol let me try it properly

**Comment by commonuserlol** [comment]
nahhh you must place inputNumber into button callback

**Comment by EduModsS** [comment]
I cut the code from layout.inputNumber and pasted it into "// your dialog code here", and it didn't work

code:

```	Menu.add(layout.button("TTTTT", () => { 		layout.inputNumber(
			"title",
			1337, {
				"fn": function(){},
				"label": "OK"
			}, {
				"fn": function(){},
				"label": "EXIT"
			}
		).then((x) => {
			x.show();
		}); }));```

**Comment by commonuserlol** [comment]
can you place .catch ?
example: `inputNumber(...).catch(e => console.log(e)).then(...)`

**Comment by EduModsS** [comment]
```
Menu.add(layout.button("TTTTT", () => {
		console.log('TTTTT CLICKED !');
		layout.inputNumber(
			"title",
			1337, {
				"fn": function(){},
				"label": "OK"
			}, {
				"fn": function(){},
				"label": "EXIT"
			}
		).catch(e => console.log(e)).then((x) => {
			x.show();
		});
	}));
```

I I tried clicking 3 times, console:

```
└╼ ❯❯❯ frida -H localhost -f com.kiloo.subwaysurf -l testjavamenu.js
     ____
    / _  |   Frida 16.2.1 - A world-class dynamic instrumentation toolkit
   | (_| |
    > _  |   Commands:
   /_/ |_|       help      -> Displays the help system
   . . . .       object?   -> Display information about 'object'
   . . . .       exit/quit -> Exit
   . . . .
   . . . .   More info at https://frida.re/docs/home/
   . . . .
   . . . .   Connected to localhost (id=socket@localhost)
Spawned `com.kiloo.subwaysurf`. Resuming main thread!
[Remote::com.kiloo.subwaysurf ]-> Menu already showed, ignoring `show` call
TTTTT CLICKED !
TTTTT CLICKED !
TTTTT CLICKED !
[Remote::com.kiloo.subwaysurf ]->
```

**Comment by commonuserlol** [comment]
Okay i just tested code and it works, it need a bit more debugging so firstly tell me:
* Your ROM name (MIUI, LineageOS, ...)
* Android version
* Output of `Menu.activityInstance.then(console.log, console.error)`

**Comment by EduModsS** [comment]
LineageOs (not official), Android 11,

**Comment by commonuserlol** [comment]
what's the output of code above? you can execute it in REPL

**Comment by EduModsS** [comment]
I didn't understand exactly, I put:

```
async function main() {

	Menu.activityInstance.then(console.log, console.error);
        ...
```

, when the menu already opened, I went back to termux and there was no console.log

**Comment by commonuserlol** [comment]
The issue probably due Frida and ROM. Try flash other rom, else activity-based actions (like dialog) won't work

**Comment by EduModsS** [comment]
now i am using another rom, the miui rom, it still hasn't solved the problem, after looking where it is frozen it is in "await Menu2.activityInstance"

**Comment by EduModsS** [comment]
Java.choose(Menu2.Api.Activity.$className, {
					onMatch: (instance) => {
						console.log("onMatch [activityInstance]");
						if (instance.getComponentName().getClassName() == Menu2.launcher) {
							resolve(Java.retain(instance));
							return "stop";
						}
					},
					onComplete() {
						console.log("onComplete [activityInstance]");
					}
				});

this code of yours, onMatch never finds anything

**Comment by commonuserlol** [comment]
What's MIUI version u have

**Comment by EduModsS** [comment]
Nevermind, the rom is Android 11, but I just asked a friend to test it and it worked, maybe it's a problem on my phone lol (he and me used the same apk modified with gadget lib)

**Comment by EduModsS** [comment]
also i am using Android 11 and he Android 13

**Comment by commonuserlol** [comment]
It's rom issue. MIUI knows to cause bugs with frida. basically this issue about frida java bridge, so I can do nothing with that

**Comment by EduModsS** [comment]
Hey bro, You could put something so that the person can specify the main activity

**Comment by EduModsS** [comment]
after doing this

```
globalThis.AppLauncher = "com.moba.unityplugin.MobaGameUnityActivity";
```
...

```
Menu2.getter(Menu2, "launcher", () => globalThis.AppLauncher ?? Menu2.app.packageManager.getLaunchIntentForPackage(Menu2.app.packageName).resolveActivityInfo(Menu2.app.packageManager, 0).name.value, Menu2.lazy);
```
```
Menu2.getter(Menu2, "activityInstance", () => {
			return new Promise((resolve, reject) => {
				Java.choose(Menu2.launcher, {
					onMatch: (instance) => {
						if (instance.getComponentName().getClassName() == Menu2.launcher) {
							resolve(Java.retain(instance));
							return "stop";
						}
					},
					onComplete() {
						
					}
				});
			});
		}, Menu2.lazy);
```



The function "layout.inputNumber and layout.inputText" is now working fine 100%
